### PR TITLE
fbo_rtt_texture_array

### PR DESCRIPTION
**Pull request by trungtle**
Using framebuffer with a 2D texture array

### CODE REVIEW

**Comment by trungtle** [comment]
@pjcozzi @shrekshao Ready for review. Currently also doesn't seem to work on Firefox OSX, but I was able to render a 2D array texture in Chrome Canary

File: README.md [hunk 1]
```diff
@@ -21,7 +21,8 @@ Inspired by and ported from Christophe Riccio's ([@Groovounet](https://github.co
 |[sampler_object](http://webglsamples.org/WebGL2Samples/#sampler_filter)|:x:D3D not supported|:white_check_mark:| :x: D3D not supported|:white_check_mark:|
 |[sampler_wrap](http://webglsamples.org/WebGL2Samples/#sampler_filter)|:white_check_mark:|:white_check_mark:| :white_check_mark:|:white_check_mark:|
 |[sampler_filter](http://webglsamples.org/WebGL2Samples/#sampler_filter)|:white_check_mark:|:white_check_mark:|:white_check_mark:|:white_check_mark:|
-|[transform_feedback_interleaved](http://webglsamples.org/WebGL2Samples/#transform_feedback_interleaved)|:white_check_mark:|:grey_question:|:white_check_mark:|:grey_question:|
+|[transform_feedback_interleaved](http://webglsamples.org/WebGL2Samples/#transform_feedback_interleaved)|:white_check_mark:|:white_check_mark:|:white_check_mark:|:white_check_mark:|
+|[transform_feedback_interleaved](http://webglsamples.org/WebGL2Samples/#transform_feedback_interleaved)|:grey_question:|:white_check_mark:|:grey_question:|:x: not working|
```
Comment (shrekshao) [inline README.md]: The name is wrong

Comment (shrekshao) [inline README.md]: Testing on Windows:
Chrome 49: [.CommandBufferContext]GL ERROR :GL_INVALID_FRAMEBUFFER_OPERATION : glDrawArraysIntancedANGLE: framebuffer incomplete (clear)

Firefox 45: Error: WebGL: framebufferTextureLayer: attachment:: invalid enum value COLOR_ATTACHMENT0 fbo_rtt_texture_array.html:255:13
Error: WebGL: drawArraysInstanced: Incomplete framebuffer: Status 0x8cdd: Bad status according to the driver.

File: samples/fbo_rtt_texture_array.html [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+
+        layout(location = POSITION_LOCATION) in vec2 position;
+        layout(location = TEXCOORD_LOCATION) in vec2 textureCoordinates;
+
+        out vec2 st;
+
+        void main()
+        {
+            st = textureCoordinates;
+            gl_Position = mvp * vec4(position, 0.0, 1.0) ;
```
Comment (shrekshao) [inline samples/fbo_rtt_texture_array.html]: catch one, extra space before ';'

File: samples/fbo_rtt_texture_array.html [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+        precision highp float;
+        precision highp int;
+
+        uniform mat4 mvp;
+
+        layout(location = POSITION_LOCATION) in vec2 position;
+
+        void main()
+        {
+            gl_Position = mvp * vec4(position, 0.0, 1.0) ;
```
Comment (shrekshao) [inline samples/fbo_rtt_texture_array.html]: extra space before ';'

File: README.md (hunk 1 above)
Comment (trungtle) [inline README.md]: I'll try to see if I could run it on a windows system. Marking as crashed for now

File: samples/fbo_rtt_texture_array.html (hunk 3 above)
Comment (trungtle) [inline samples/fbo_rtt_texture_array.html]: Fixed

File: samples/fbo_rtt_texture_array.html (hunk 2 above)
Comment (trungtle) [inline samples/fbo_rtt_texture_array.html]: Fixed

**Comment by trungtle** [comment]
This pull request is pending due to issue on Windows still crashing, I haven't had time to investigate further. Will look more into it.

**Comment by pjcozzi** [comment]
In Canary on Windows, I get:

```
[.CommandBufferContext]GL ERROR :GL_INVALID_FRAMEBUFFER_OPERATION : glDrawArraysIntancedANGLE: framebuffer incomplete (clear)
```

Are you sure the framebuffer is complete?  There are a lot of rules regarding what texture formats are compatible when bound to the same framebuffer.

**Comment by trungtle** [comment]
I don't currently have that problem on Chrome OSX, but I did see this replicate in Chrome Windows. And also Firefox (in both Windows and OSX) produces 

```
Firefox 45: Error: WebGL: framebufferTextureLayer: attachment:: invalid enum value COLOR_ATTACHMENT0 fbo_rtt_texture_array.html:255:13
Error: WebGL: drawArraysInstanced: Incomplete framebuffer: Status 0x8cdd: Bad status according to the driver.
```

So I think it definitely has something to do with the framebuffer being incomplete

**Comment by pjcozzi** [comment]
Use `checkFramebufferStatus` to check for completion.

See [Section 4.4.4 Framebuffer Completeness](https://www.khronos.org/registry/gles/specs/3.0/es_spec_3.0.4.pdf) of the OpenGL ES 3.0 spec.

Also see the WebGL 2 spec:

> 5.27 checkFramebufferStatus may return FRAMEBUFFER_INCOMPLETE_DIMENSIONS
> 
> All attached images much have the same width and height; otherwise, checkFramebufferStatus returns FRAMEBUFFER_INCOMPLETE_DIMENSIONS.
> 
> In OpenGL ES 3, attached images for a framebuffer need not to have the same width and height to be framebuffer complete, and FRAMEBUFFER_INCOMPLETE_DIMENSIONS is not one of the valid return values for checkFramebufferStatus. However, in Direct3D 11, on top of which OpenGL ES 3 behavior is emulated in Windows, all render targets must have the same size in all dimensions (see msdn manual page). Emulation of the ES3 semantic on top of DirectX 11 will be inefficient. In order to have consistent WebGL 2 behaviors across platforms, it is reasonable to keep the OpenGL ES 2 / WebGL 1 restriction for WebGL 2 that all attached images must have the same width and height.

**Comment by trungtle** [comment]
@pjcozzi This problem only reproduces on Windows and Firefox.

I am using a texture array with the same with and height, and this check below:

```
gl.checkFramebufferStatus(gl.FRAMEBUFFER) == gl.FRAMEBUFFER_COMPLETE
```

also returns true when I check it after the drawBuffers() and drawArraysInstanced(). Additionally, Firefox complains about invalid enum gl.COLOR_ATTACHMENT0, but not COLOR_ATTACHMENT1 or COLOR_ATTACHMENT2, I suspect it's a Firefox implementation issue.

So far, only Chrome OSX gives me no error, so I wonder if I should wrap this up, merge, and open a ticket for this?

**Comment by pjcozzi** [comment]
Try it in Chrome on Windows with ANGLE disabled.

**Comment by kenrussell** [comment]
It would be easiest for us to reproduce and debug the failures if this pull request were merged -- or if another reduced test case were created and a bug filed on crbug.com/ .

**Comment by trungtle** [comment]
@kenrussell Thank you for looking into this! I think this is ready to merge, I'm waiting to see what Patrick thinks. 

@pjcozzi. Whenever I disabled ANGLE in Chrome (using the --use-gl=desktop flag, if that's the right way), I kept crashing the browser. What should happen if ANGLE is disabled when we're expected to run WebGL2? Does it try to use some fallback to run Direct3D or would it crash (like what I have seen)

**Comment by kenrussell** [comment]
--use-gl=desktop takes an older and unmaintained code path in Chrome. Try --use-angle=gl instead. It may also currently require --disable-gpu-sandbox. Don't use these flags for browsing the general Web.

**Comment by pjcozzi** [comment]
Looks like it works in Canary on Windows without ANGLE (I ran with `--enable-unsafe-es3-apis --use-angle=gl --disable-gpu-sandbox`):

image

Will merge this now to help @kenrussell and crew with testing.

**Comment by kenrussell** [comment]
To be clear, that's still using ANGLE, but with its OpenGL backend rather than its D3D backend.

**Comment by pjcozzi** [comment]
Ah, yes, I see.

**Pull request merged**